Tests/SonarScanner.MSBuild.PreProcessor.Test/JreCaching/JreCacheTests.cs

```diff
@@ -538,6 +553,295 @@ public async Task Checksum_DownloadFile_FileOpenFails()
+        fileWrapper.Received(1).Exists(file);
+        fileWrapper.Received(1).Create(Arg.Any<string>());
+        fileWrapper.Received(2).Open(file); // One for the checksum and the other for the unpacking.
+        checksum.Received(1).ComputeHash(Arg.Any<Stream>());
+        unpackerFactory.Received(1).CreateForArchive(directoryWrapper, fileWrapper, TestArchiveName);
+        testLogger.DebugMessages.Should().SatisfyRespectively(
+            x => x.Should().Be(@"Starting the Java Runtime Environment download."),
+            x => x.Should().Be(@"The checksum of the downloaded file is 'sha256' and the expected checksum is 'sha256'."),
+            // The unpackerFactory returned an unpacker and it was called. But the test setup is incomplete and therefore fails later:
+            x => x.Should().StartWith(@"Starting extracting the Java runtime environment from archive 'C:\Users\user\.sonar\cache\sha256\filename.tar.gz' to folder 'C:\Users\user\.sonar\cache\sha256"),
+            x => x.Should().StartWith(@"The extraction of the downloaded Java runtime environment failed with error 'The java executable in the extracted Java runtime environment was expected to be at 'C:\Users\user\.sonar\cache\sha256"));
+    }
+
+    [TestMethod]
+    public async Task UnpackerFactory_ReturnsNull()
+    {
+        var home = @"C:\Users\user\.sonar";
+        var cache = Path.Combine(home, "cache");
+        var sha = Path.Combine(cache, "sha256");
+        directoryWrapper.Exists(cache).Returns(true);
+        directoryWrapper.Exists(sha).Returns(true);
+        unpackerFactory.CreateForArchive(directoryWrapper, fileWrapper, TestArchiveName).ReturnsNull();
+
+        var sut = CreateSutWithSubstitutes();
+        var result = await sut.DownloadJreAsync(home, new(TestArchiveName, "sha256", "javaPath"), () => Task.FromResult<Stream>(new MemoryStream()));
+        result.Should().BeOfType<JreCacheFailure>().Which.Message.Should().Be("The archive format of the JRE archive `filename.tar.gz` is not supported.");
+        fileWrapper.DidNotReceiveWithAnyArgs().Exists(null);
+        fileWrapper.DidNotReceiveWithAnyArgs().Create(null);
+        fileWrapper.DidNotReceiveWithAnyArgs().Open(null);
+        checksum.DidNotReceiveWithAnyArgs().ComputeHash(null);
+        unpackerFactory.Received(1).CreateForArchive(directoryWrapper, fileWrapper, TestArchiveName);
+        testLogger.DebugMessages.Should().BeEmpty();
+    }
+
+    [TestMethod]
+    public async Task UnpackerFactory_Throws()
+    {
+        var home = @"C:\Users\user\.sonar";
+        var cache = Path.Combine(home, "cache");
+        var sha = Path.Combine(cache, "sha256");
+        directoryWrapper.Exists(cache).Returns(true);
+        directoryWrapper.Exists(sha).Returns(true);
+        unpackerFactory.CreateForArchive(directoryWrapper, fileWrapper, TestArchiveName).ReturnsNull();
```

Comment:
Shouldn't this throw instead of return null?

Reply:
That is what it did before, but you requested to change it: https://github.com/SonarSource/sonar-scanner-msbuild/pull/2018#discussion_r1668511430

Reply:
Oh, yeah, sry. The test name is not up to date and it confused me.

Reply:
I corrected the test name.